File: Core/YABACore/src/commonMain/kotlin/dev/subfly/yabacore/model/annotation/AnnotationSourceContext.kt
package dev.subfly.yabacore.model.annotation

import androidx.compose.runtime.Stable
import kotlinx.serialization.Serializable

/**
 * Identifies which bookmark/content/version an annotation belongs to.
 */
@Serializable
@Stable
data class AnnotationSourceContext(
    val bookmarkId: String,
    val type: AnnotationType,
    /** For READABLE: readableVersionId. For PDF: same id space as stored on the entity. */
    val contentId: String,
) {
    companion object {
        fun readable(bookmarkId: String, readableVersionId: String): AnnotationSourceContext =
            AnnotationSourceContext(
                bookmarkId = bookmarkId,
                type = AnnotationType.READABLE,
                contentId = readableVersionId,
            )

        fun pdf(bookmarkId: String, readableVersionId: String): AnnotationSourceContext =
            AnnotationSourceContext(
                bookmarkId = bookmarkId,
                type = AnnotationType.PDF,
                contentId = readableVersionId,
            )

        fun epub(bookmarkId: String, readableVersionId: String): AnnotationSourceContext =
            AnnotationSourceContext(
                bookmarkId = bookmarkId,
                type = AnnotationType.EPUB,
                contentId = readableVersionId,
            )
    }
}

File: Darwin/YABACore/Model/Annotation/AnnotationSourceContext.swift
//
//  AnnotationSourceContext.swift
//  YABACore
//
//  Parity with Compose `AnnotationSourceContext`.
//

import Foundation

public struct AnnotationSourceContext: Sendable, Codable, Equatable {
    public var bookmarkId: String
    public var type: AnnotationType

    public init(bookmarkId: String, type: AnnotationType) {
        self.bookmarkId = bookmarkId
        self.type = type
    }

    public static func readable(bookmarkId: String) -> AnnotationSourceContext {
        AnnotationSourceContext(bookmarkId: bookmarkId, type: .readable)
    }

    public static func pdf(bookmarkId: String) -> AnnotationSourceContext {
        AnnotationSourceContext(bookmarkId: bookmarkId, type: .pdf)
    }

    public static func epub(bookmarkId: String) -> AnnotationSourceContext {
        AnnotationSourceContext(bookmarkId: bookmarkId, type: .epub)
    }
}
